dll/win32/iernonce/dialog.cpp

```diff
@@ -0,0 +1,220 @@
+/*
+ * PROJECT:     ReactOS system libraries
+ * LICENSE:     GPL-2.0-or-later (https://spdx.org/licenses/GPL-2.0-or-later)
+ * PURPOSE:     Classes for displaying progress dialog.
+ * COPYRIGHT:   Copyright 2021 He Yang <1160386205@qq.com>
+ */
+
+
+#include <atlbase.h>
+#include <atlwin.h>
+#include <windowsx.h>
+#include <process.h>
+
+#include "dialog.h"
+#include "registry.h"
+#include "thread.h"
+
+#define ITEM_VPADDING     3
+#define ITEM_LEFTPADDING 22
+
+HFONT CreateBoldFont(HFONT hOrigFont)
+{
+    LOGFONT fontAttributes = { 0 };
+    GetObject(hOrigFont, sizeof(fontAttributes), &fontAttributes);
+    fontAttributes.lfWeight = FW_BOLD;
+
+    return CreateFontIndirect(&fontAttributes);
+}
+
+ProgressDlg::ProgressDlg(RunOnceExInstance &RunOnceExInst) :
+    m_DialogID(-1), m_hListBox(NULL), m_hBoldFont(NULL), m_PointedItem(0), m_RunOnceExInst(RunOnceExInst)
+{ ; }
+
+BOOL ProgressDlg::RunDialogBox()
+{
+    // only when the list is not empty.
+    if (m_RunOnceExInst.m_SectionList.GetSize() != 0)
+    {
+        m_DialogID = DoModal();
+        if (m_DialogID == -1)
+        {
+            return FALSE;
+        }
+    }
+    return TRUE;
+}
+
+void ProgressDlg::CalcTextRect(LPCWSTR lpText, RECT *pRect)
+{
+    HDC hdc = ::GetDC(m_hListBox);
+    ::GetClientRect(m_hListBox, pRect);
+
+    pRect->bottom = pRect->top;
+    pRect->left += ITEM_LEFTPADDING;
+
+    HFONT OldFont = SelectFont(hdc, GetFont());
+
+    DrawTextW(hdc, lpText, -1, pRect, DT_CALCRECT | DT_WORDBREAK);
+    ::ReleaseDC(m_hListBox, hdc);
+
+    SelectFont(hdc, OldFont);
```

Comment:
you release the DC, but after that you change the font?

Reply:
Oh, I'm trying to restore to the origin font 🤣
I found probably it's no need to do this, As this hdc is temporary?

Reply:
as soon as the releasedc has been called, the hdc becomes invalid, so you should just restore the font before releasing the dc!

Reply:
This DC is just for calculate the text height 🤣, I think there's no need to restore the old font...?

Reply:
by having set an new font in the hdc you effectively changed it, so if you don't restore the old one you're good for some surprises.

Reply:
I always thought after the DC is released everything that has been selected into it will no longer taking effect! 😥
Ok then, I will fix it √